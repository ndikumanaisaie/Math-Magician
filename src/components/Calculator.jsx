/* eslint-disable no-nested-ternary */
/* eslint-disable react/prefer-stateless-function */
import React, { Component } from 'react';

class Calculator extends Component {
  render() {
    const btnValues = [
      ['AC', '+-', '%', 'รท'],
      [7, 8, 9, 'x'],
      [4, 5, 6, '-'],
      [1, 2, 3, '+'],
      [0, '.', '='],
    ];

    return (
      <div className="wrapper">
        <div className="screen">0</div>
        <div className="button-box">
          {
            btnValues.flat().map((btn) => (
              <button
                key={btn}
                type="button"
                className={
                  btn === '=' || btn === 'รท' || btn === 'x' || btn === '-' || btn === '+'
                    ? 'sign-btn'
                    : btn === 0
                      ? 'zero'
                      : '0'
                }
                value={btn}
              >
                { btn }
              </button>
            ))
          }
        </div>
      </div>
    );
  }
}

export default Calculator;
