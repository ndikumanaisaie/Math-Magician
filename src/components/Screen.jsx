/* eslint-disable react/prefer-stateless-function */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Screen extends Component {
  render() {
    const { value } = this.props;
    return (
      <div className="screen">{value}</div>
    );
  }
}

Screen.propTypes = {
  value: PropTypes.string.isRequired,
};

export default Screen;
