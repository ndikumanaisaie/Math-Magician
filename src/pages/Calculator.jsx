/* eslint-disable no-nested-ternary */
/* eslint-disable react/prefer-stateless-function */
import React, { useState } from 'react';
import btnValues from '../utilities/util';
import Screen from '../components/Screen';
import calculate from '../logic/calculate';

const Calculator = () => {
  const [calc, setCalc] = useState({
    operation: null,
    next: null,
    total: null,
  });

  const { operation, next, total } = calc;
  const handleClick = (e) => {
    setCalc(calculate({ total, next, operation }, e.target.value));
  };

  return (
    <div className="main-wrapper">
      <h2 className="calc-title">
        Let&apos;s do some math
      </h2>
      <div className="wrapper">
        <Screen value={next || total || '0'} />
        <div className="button-box">
          {
            btnValues.flat().map((btn) => (
              <button
                key={btn}
                type="button"
                className={
                  btn === '=' || btn === 'รท' || btn === 'x' || btn === '-' || btn === '+'
                    ? 'sign-btn'
                    : btn === 0
                      ? 'zero'
                      : '0'
                }
                value={btn}
                onClick={handleClick}
              >
                { btn }
              </button>
            ))
          }
        </div>
      </div>
    </div>

  );
};

export default Calculator;
