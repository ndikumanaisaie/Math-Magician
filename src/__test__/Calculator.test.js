import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Screen from '../components/Screen';

import Calculator from '../pages/Calculator';

describe('Calculator page component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly', () => {
    const screenTree = renderer.create(<Screen />).toJSON();
    expect(screenTree).toMatchSnapshot();
  });

  it('should return the multiplication of two number from UI', () => {
    render(<Calculator />);
    fireEvent.click(screen.getByText('5'));
    fireEvent.click(screen.getByText('x'));
    fireEvent.click(screen.getByText('7'));
    fireEvent.click(screen.getByText('='));
    expect(screen.getByText(/35/)).toBeInTheDocument();
  });
});
