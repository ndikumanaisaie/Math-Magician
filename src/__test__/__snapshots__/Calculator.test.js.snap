// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator page component renders correctly 1`] = `
<div
  className="main-wrapper"
>
  <h2
    className="calc-title"
  >
    Let's do some math
  </h2>
  <div
    className="wrapper"
  >
    <div
      className="screen"
    >
      0
    </div>
    <div
      className="button-box"
    >
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value="AC"
      >
        AC
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value="+-"
      >
        +-
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value="%"
      >
        %
      </button>
      <button
        className="sign-btn"
        onClick={[Function]}
        type="button"
        value="รท"
      >
        รท
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value={7}
      >
        7
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value={8}
      >
        8
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value={9}
      >
        9
      </button>
      <button
        className="sign-btn"
        onClick={[Function]}
        type="button"
        value="x"
      >
        x
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value={4}
      >
        4
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value={5}
      >
        5
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value={6}
      >
        6
      </button>
      <button
        className="sign-btn"
        onClick={[Function]}
        type="button"
        value="-"
      >
        -
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value={1}
      >
        1
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value={2}
      >
        2
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value={3}
      >
        3
      </button>
      <button
        className="sign-btn"
        onClick={[Function]}
        type="button"
        value="+"
      >
        +
      </button>
      <button
        className="zero"
        onClick={[Function]}
        type="button"
        value={0}
      >
        0
      </button>
      <button
        className="0"
        onClick={[Function]}
        type="button"
        value="."
      >
        .
      </button>
      <button
        className="sign-btn"
        onClick={[Function]}
        type="button"
        value="="
      >
        =
      </button>
    </div>
  </div>
</div>
`;

exports[`Calculator page component renders correctly 2`] = `
<div
  className="screen"
>
  0
</div>
`;
